Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,509 @@
+class ConnectionPool {
+    /// The default look group used to schedule async work
+    let loopGroup: EventLoopGroup
+
+    /// The configuration used to bootstrap new HTTP connections
+    let configuration: HTTPClient.Configuration
+
+    /// The main data structure used by the `ConnectionPool` to retreive and create connections associated
+    /// to a a given `Key` .
+    /// - Warning: This property shouldn't be directly accessed, use the `ConnectionPool` subscript instead
+    private var _connectionProviders: [Key: ConnectionProvider] = [:]
+
+    private let lock = Lock()
+
+    init(group: EventLoopGroup, configuration: HTTPClient.Configuration) {
+        self.loopGroup = group
+        self.configuration = configuration
+    }
+
+    /// This enables the connection pool to store and retreive `ConnectionProvider`s
+    /// by ensuring thread safety using the `lock`.
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return self.lock.withLock {
+                _connectionProviders[key]
+            }
+        }
+        set {
+            self.lock.withLock {
+                _connectionProviders[key] = newValue
+            }
+        }
+    }
+
+    /// This method asks the pool for a connection usable by the specified `request`, respecting the specified options.
+    ///
+    /// - parameter request: The request that needs a `Connection`
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        return self.lock.withLock {
+            let key: Key
+            // Created a key and checks the URL components are valid
+            do {
+                key = try Key(url: request.url)
+            } catch {
+                return self.loopGroup.next().makeFailedFuture(error)
+            }
+
+            if let provider = self._connectionProviders[key] {
+                // If a connection provider already exists for the given key,
+                // directly ask it for a new connection
+                return provider.getConnection(preference: preference)
+            } else {
+                // If no connection provider exists for the given key, create a new one
```

Comment:
Hey, I have one question: Isn't this leading to a race condition? If I fire two or more requests for the same `Key` the first one will not find a connection provider, so it will create a new connection. If the second request comes in before the first connection was saved into the dictionary in line 131, then the second request will also create a new provider and save it later to the dictionary. Which of the two requests will win is undefined. Is this intended behaviour?

Reply:
That seems not ideal. I think a "tombstone" (or even a promise) could be used to resolve this issue.

Reply:
@t089  Hey, thanks I think you are right and now that I think about it this might be linked to another random issue that sometimes happened where a `ConnectionProvider` seemed to be deinitialized too early. More generally I'm going to try to limit the use of locks (and at least avoid acquiring locks other such large portions of the code as its currently the case) and also use other mechanisms as suggested by @Lukasa.

Reply:
@t089 @Lukasa The https://github.com/swift-server/async-http-client/pull/105/commits/f3a84bf4adf846048adbf1c02b4e9d9f714c675e commit attempts to solve this issue. WDYT?

Reply:
Looks a lot better to me.